### ISSUE TITLE
Change the prototype of functions that deal with backup GPT?

### ISSUE THREAD

**Issue by lurch** [jhermsmeier/node-gpt#9]
Currently, the functions that deal with backup GPTs (i.e. `parseBackup` and `writeBackup`) have a function-prototype which accepts a `buffer` and an `offset`.

The `parseBackup` function reads the table-entries at `offset` and the header at `buffer.length - this.blockSize` - however this seems to assume that `offset` is set to _exactly_ the right amount of distance before the end of the `buffer`, which feels incredibly fragile?

The `writeBackup` function writes the table-entries at `offset` and the header at `offset + tableSize` which is a subtle (but IMHO fundamentally different) method of operation from `parseBackup`. Again this is fragile as it assumes that the `offset` is positioned _just_ right, and that the `buffer` is big enough to hold `offset + tableSize + blockSize`.

From reading the GPT specs, it feels (to me) like the `offset` passed into both the `parseBackup` and `writeBackup` functions _ought_ to be the `headerOffset` of the backup GPT header (i.e. which is typically the last sector on the disk). However this assumes that the `buffer` being passed in is `seek()`-able - is that a safe assumption to make?
When reading the backupGPT, you need to _first_ read the "Starting LBA of array of partition entries" from the header in order to calculate how far backwards to seek to begin reading partition-entries from, and when writing the backupGPT you need to know how far backwards you need to seek so that you start writing the partition-entries at "the right place".

However these changes would obviously result in a change-of-API, and so would necessitate a major version-bump.

As an example of why this matters: imagine that you have a disk with a totally corrupt primaryGPT but a valid backupGPT - with the way that `node-gpt` currently works you'd have no idea what `offset` value to pass to `parseBackup`. (remember that GPT allows an unlimited number of partitions, so the partition-table-array _may_ be larger than 16KB)

I don't mean to sound picky, I'm just trying to make the code more robust :slightly_smiling_face:

**Comment by lurch** [comment]
Also, I wonder if modifying the code so that LBAs were passed around (instead of byte-offsets being passed around) would make it more readable and easier to tally-up with the GPT specs? :shrug:

**Comment by jhermsmeier** [comment]
> Also, I wonder if modifying the code so that LBAs were passed around (instead of byte-offsets being passed around) would make it more readable and easier to tally-up with the GPT specs?

No, because you'd have to translate back to byte offsets for every operation on buffers – makes the code noisy as hell and far less readable. Also easier to debug with byte offsets.

**Comment by jhermsmeier** [comment]
> The `parseBackup` function reads the table-entries at `offset` and the header at `buffer.length - this.blockSize` - however this seems to assume that `offset` is set to _exactly_ the right amount of distance before the end of the `buffer`, which feels incredibly fragile?

It's the other way around, offset is zero by default. The header is read from the end of the buffer because the order of header & table is flipped for the secondary GPT.
It will fail if you pass in a buffer that does not contain the GPT header at the end of it, in which case it should fail.

**Comment by jhermsmeier** [comment]
> The `writeBackup` function writes the table-entries at `offset` and the header at `offset + tableSize` which is a subtle (but IMHO fundamentally different) method of operation from `parseBackup`. Again this is fragile as it assumes that the `offset` is positioned _just_ right, and that the `buffer` is big enough to hold `offset + tableSize + blockSize`.

Same as above here; offset starts at the beginning, not end. This is supposed to break if the buffer doesn't fit table + header. These methods are not supposed to operate on random inputs.

**Comment by jhermsmeier** [comment]
> However this assumes that the `buffer` being passed in is `seek()`-able - is that a safe assumption to make?

https://nodejs.org/api/buffer.html

**Comment by jhermsmeier** [comment]
> As an example of why this matters: imagine that you have a disk with a totally corrupt primaryGPT but a valid backupGPT - with the way that `node-gpt` currently works you'd have no idea what `offset` value to pass to `parseBackup`. (remember that GPT allows an unlimited number of partitions, so the partition-table-array _may_ be larger than 16KB)

Generally `.parseBackup()` is a convenience method, for situations where you already kind of know what to expect.
If you have to deal with this in a real-world situation, you shouldn't be using `.parseBackup()`, but be reading the header and checking it with `.parseHeader()`, after which you could somewhat safely use `.parseBackup()` with an offset calculated from the parsed secondary header. Ideally you'd `.parseTable()` though.

The stupid thing is that `.parseBackup()` doesn't take into account the table size & entry count properly. Should be sanity checking those against the offsets. Prob. same for `.parseBackup()`.

Should prob. deprecate both methods, and resurface as `.{read,write}Secondary()` or so, which then would do a bit of hand-holding around that.

**Comment by lurch** [comment]
> These methods are not supposed to operate on random inputs.

Sorry, didn't realise that.
